### PR TITLE
[installer] Move `personalAccessTokenSigningKey` out of experimental

### LINKED ISSUE #15028: [installer] mandate and move `personalAccessTokenSigningKeySecretName` from experimental
## Is your feature request related to a problem? Please describe

With `public-api` becoming a mandatory component and the signing key being used in `server`,
It makes sense to mandate it as a pre-requisite to be created externally and the secret name
to be passed to the installer.

## Describe the behaviour you'd like

After this change, The installer should follow the convention of mandating
a secret name to be passed for the `render` to work. This secret should then
be loaded and used into components like `public-api` and `server`. The render
should fail otherwise.

This change involves multiple things:
- [ ] mandate and move `personalAccessTokenSigningKeySecretName` from experimental
- [ ] Update `local-preview` to auto-generate a signing key and to use it.
- [ ] (Needed only if there is a `kots` release) Update `kots-install.sh` to auto-generate the signing key

## Describe alternatives you've considered

- Auto-generating the signing key does not make sense as it needs to be same across roll-outs.
- Default static key also does not make sense due to the security concerns.

## Additional context

- https://gitpod.slack.com/archives/C01KLC56NP7/p1669038281983829

### PR DESCRIPTION
**Pull request by Pothulapati**
## Description

This PR moves `personalAccessTokenSigningKey` out of `experimental`
and into `v1.Config` while marking it as `required` by default, with
a default secret name of `personal-access-token-signing-key`. It's
the environment's responsibility to make sure this secret exists for
Gitpod to work. (See #15028 and slack thread for more info on this).

Signed-off-by: Tarun Pothulapati <tarun@gitpod.io>

## Related Issue(s)

Fixes #15028

## Release Notes

```release-note
[installer] Move `personalAccessTokenSigningKey` out of experimental
```

## Werft options:

- [x] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [ ] /werft with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`

### CODE REVIEW

File: install/installer/pkg/components/public-api-server/configmap.go
Comment (Pothulapati) [inline install/installer/pkg/components/public-api-server/configmap.go:125]: Marked the secret load as mandatory now! 👀

**Review by vulkoingim — APPROVED** [review]

**Comment by corneliusludmann** [comment]
Would we consider this config change as a breaking change? If yes, what are the implications for (internal/external) users? See also https://github.com/gitpod-io/gitpod/issues/15121.

**Comment by Pothulapati** [comment]
@corneliusludmann Going by https://github.com/gitpod-io/gitpod/issues/15121#issuecomment-1335115146, I wouldn't consider this a breaking change but will wait for other's opinions before I merge this! 👍🏼

**Comment by easyCZ** [comment]
Thanks. I'll take a look this afternoon. At first glance it looks okay but want to go through it in more detail.

File: install/installer/pkg/config/v1/config.go
Comment (easyCZ) [inline install/installer/pkg/config/v1/config.go:178]: How do we communicate what properties must exist on the `SecretRef`? Here we say that this secret must exist, but we also require that the secret also contains a particular key, which has a value of the PAT Signing Key

Comment (Pothulapati) [reply install/installer/pkg/config/v1/config.go:178]: >Here we say that this secret must exist

I don't think we communicate this. We expect the field `personalAccessTokenSigningKey` to be set to some value (there is no run-time verification whether it's there) and considering we have a [default value for this](https://github.com/gitpod-io/gitpod/pull/15059/files#diff-ed0dab0d7e4e73c6487f629ed17403feeb8ee3b8cf1601e5c940f7a9365ccf17R58), This will never fail. The deployment will eventually fail if the secret does not exist as it's no longer optional.

The reason for having a default value is not to change the existing behavior as the new installer would fail wherever this field is not set 🤔  

> we also require that the secret also contains a particular key, which has a value of the PAT Signing Key

This was easy for certificates as there is a standard format for storing certs in Kubernetes secret. I don't think we have a straightforward way of doing this. We can add a comment and [render it in the docs](https://github.com/gitpod-io/gitpod/blob/main/install/installer/docs/overview.md#config) on how we expect this to be present.

cc: @mrsimonemms WDYT? 🤔

**Review by mrsimonemms — APPROVED** [review]
/hold

I'm broadly happy with this PR. I echo the comments that this is **NOT** a breaking change (although is a change that could break some installations). I take this in the same way as #13126 - anyone who wasn't doing this already had an unsupported installation and we have merely brought the validation up to date.

I would favour adding some [validation](https://github.com/gitpod-io/gitpod/pull/15059#discussion_r1039647330) (either now, or in a future PR) and I definitely like the "fail fast" approach if there is not PAT secret.

Before merging, I'd also recommend checking that there's no updates to the golden files from `main`.

Good work @tarun - I'll approve it now so you're not blocked and you can remove the hold at will

**Comment by mrsimonemms** [comment]
@Pothulapati it might also be worth also deprecating the experimental config parameter - see [code](https://github.com/gitpod-io/gitpod/blob/797c1b74214414522c01e1c7475353bf7c3812aa/install/installer/pkg/config/v1/config.go#L121)

**Comment by Pothulapati** [comment]
@mrsimonemms Makes sense. Will add the validation checks along with field deprecation.

 >I would favour adding some https://github.com/gitpod-io/gitpod/pull/15059#discussion_r1039647330 (either now, or in a future PR) and I definitely like the "fail fast" approach if there is not PAT secret.

Do you think the current methodology i.e mandatory config with default value (where the failure happens with the deployment) is fast enough? We could remove the default value even and make it fail even earlier (i.e at render if they don't set it to some value) 🤔

File: install/installer/pkg/config/v1/validation.go
Comment (Pothulapati) [inline install/installer/pkg/config/v1/validation.go:128]: @easyCZ Added validation for both the secret and key (Looking at the format in the[ preview environments file](https://github.com/gitpod-io/gitpod/blob/tar/installer-mandate-15028/.werft/jobs/build/public-api/personal-access-token-signing-key.yaml#L3)). Do you see a reason to rename the key? 🤔

File: install/installer/pkg/components/public-api-server/configmap.go
Comment (easyCZ) [inline install/installer/pkg/components/public-api-server/configmap.go:125]: You can remove the Optional field entirely

```suggestion
```

File: install/installer/pkg/config/v1/experimental/experimental.go
Comment (easyCZ) [inline install/installer/pkg/config/v1/experimental/experimental.go:268]: 👍

File: install/installer/pkg/config/v1/validation.go
Comment (easyCZ) [reply install/installer/pkg/config/v1/validation.go:128]: Not really. I think we can stick with this name for now.

**Review by easyCZ — APPROVED** [review]
Looks good. Before we land this, we should make corresponding changes to the `ops` config such that when we roll this out, it will already have the reference specifeid. I can do that shortly.

/hold for ops change

**Comment by mrsimonemms** [comment]
> Do you think the current methodology i.e mandatory config with default value (where the failure happens with the deployment) is fast enough? We could remove the default value even and make it fail even earlier (i.e at render if they don't set it to some value) thinking

@Pothulapati yeah, I would make the `render` error if there is no PAT secret. The `required` validation field should be enough as the validation is run as part of the `render` command - yes, there's a way around it (using `--no-validation`), but it's how we've done it with everything else.

I guess you could return an error as well if there is no PAT secret in the reference as well, but I would only do that if it'll return a `nil` pointer error otherwise as they're a pain in the backside to debug

**Comment by Pothulapati** [comment]
Updated the PR to make it mandated, without a default value like we discussed. Marking the experimental value deprecated also means that setting it will still work, but renders a warning! 👀

**Comment by Pothulapati** [comment]
@easyCZ I've tested that using `experimental.webapp.publicAPI.personalAccessTokenSigningKeySecretName` field to set the value will continue to work, but throws the following warning:
```
Deprecated config parameter: experimental.webapp.publicAPI.personalAccessTokenSigningKeySecretName=some-override
```

I think, This PR is ready to be merged now. Are you planning to raise the `ops` PR? I can take it up if you are busy! 👀

**Comment by easyCZ** [comment]
@Pothulapati The config PR is here https://github.com/gitpod-io/ops/pull/7354

**Comment by Pothulapati** [comment]
Merging this then!

**Comment by easyCZ** [comment]
The config needs to land first, otherwise if we rolled out we'd not have the matching config in place

**Comment by corneliusludmann** [comment]
> Updated the PR to make it mandated, without a default value like we discussed.

I don't know if I missed the discussion in this PR somewhere. As far as I see, that is (besides `domain`) the first config value that is mandatory and has no default value. That's IMO the reason that we never thought about whether this case would be a breaking change or not (I more and more believe it is one but that is something that we should discuss in https://github.com/gitpod-io/gitpod/pull/15189). We never introduced a mandatory config value without default.

What is the reason that we are doing this now? Unless there is a good reason, I would propose doing the same as with [HTTPS certificate](https://github.com/gitpod-io/gitpod/blob/4e445233a740d711aafb1bc6a20c9a6eb4b3c801/install/installer/pkg/config/v1/config.go#L55-L56): Set a default secret name in the defaults and let the user override the default in the config. Then we will stick with the principle that you only need to configure the domain and for everything else, we have sensible defaults.

Or do we want to refrain from the principle? Do we want to remove more defaults and force the user to specify them explicitly? When there is a good reason for this, we should also document and justify this in https://github.com/gitpod-io/gitpod/pull/15189.

To unblock Team WebApp, I would be okay with merging this now. However, I think we need to clarify this very soon, document our decision and if necessary, straighten it out in a follow-up PR as soon as possible.

What do you think, @Pothulapati and @mrsimonemms?

**Comment by Pothulapati** [comment]
>What is the reason that we are doing this now? Unless there is a good reason, I would propose doing the same as with [HTTPS certificate](https://github.com/gitpod-io/gitpod/blob/4e445233a740d711aafb1bc6a20c9a6eb4b3c801/install/installer/pkg/config/v1/config.go#L55-L56):

For `https-certificates`, We use `cert-manager` as a dependency which generates it for us with the same name. This works because we specify `cert-manager` as a mandatory dependency and this workflow hence works.

For `personal-access-token-signing-key` We don't have an external component that installer asks to generate it. This means without us failing at render (i.e not setting a default value), The deployment (i.e webapp component) would eventually fail with the lack of this secret present, and would be confusing to the user. Me and @mrsimonemms discussed the same and felt it would be better to fail earlier than later, along with validations for the same. 🤔

**Comment by corneliusludmann** [comment]
> For `https-certificates`, We use `cert-manager` as a dependency which generates it for us with the same name. This works because we specify `cert-manager` as a mandatory dependency and this workflow hence works.
> 
> For `personal-access-token-signing-key` We don't have an external component that installer asks to generate it. This means without us failing at render (i.e not setting a default value), The deployment (i.e webapp component) would eventually fail with the lack of this secret present, and would be confusing to the user.

Thanks for the explanation, Tarun! :+1: 

I absolutely agree with the principle to fail early. However, I don't get why there is a difference between the personal access token secret and the HTTPS certs secret. Especially since the existence of this config value does not give any guarantees that this secret actually exists. The dependency on the `https-certificates` secret is based solely on the [docs where we say there must be secret with the same name as the value in the config](https://github.com/gitpod-io/gitpod/blob/4e445233a740d711aafb1bc6a20c9a6eb4b3c801/install/installer/docs/overview.md#tls-certificates). One _possible_ way to get the secret is with the `cert-manager` [where we use the secret name in the sample code in the docs](https://github.com/gitpod-io/gitpod/blob/4e445233a740d711aafb1bc6a20c9a6eb4b3c801/install/installer/docs/overview.md#2-create-the-certificate). However, there is no guarantee that the user is configuring cert-manager the same or use a completely different way to provide the HTTPS certs as a secret with another name. And the cert-manager is a mandatory dependency for internal certs we use, not for the public HTTPS certs. It's totally up to the user how they create the secret (and which name they use).

The only reason I can see currently for making the config value mandatory for the user is that they are _aware_ of the need for a new secret when they _upgrade_ an existing Gitpod installation. When they simply upgrade without reading the docs, it fails (early!). However, this is a breaking change of the requirements we have (which is checked with the cluster validate command that should be better run before every update) and not a breaking change in the config.

But I'm happy to be convinced of a different view if you see it differently.

**Comment by Pothulapati** [comment]
Discussed offline about this with @corneliusludmann, and the plan is to add a default value for this newly added field which means its a behavioural breaking change and not a config breaking change.

Will update the PR once the above change is made!

File: install/installer/cmd/testdata/render/pat-key/config.yaml
Comment (corneliusludmann) [inline install/installer/cmd/testdata/render/pat-key/config.yaml:5]: What is the purpose of this test? It's the same as the `minimal` test, right?

Is this …

1. a left-over, or
2. a test where we should add a custom value for the `personalAccessTokenSigningKey` config?

**Review by corneliusludmann — APPROVED** [review]
Looks good to me. Added a comment regarding the `pat-key` test. I'll keep the hold label so that you can address this comment (and also due to the holiday chill period).

**Comment by geropl** [comment]
@corneliusludmann Closing the PR for now, as I don't see any activity and it sits in our queue atm.

Happy to re-open any time! :bow:

**Pull request closed without merging**
